### PR TITLE
Feature/mct optim fix

### PR DESCRIPTION
**Pull request by corentincarton**

### CODE REVIEW

File: src/lisflood/hydrological_modules/kinematic_wave_parallel_tools.py [hunk 1]
```diff
@@ -60,91 +62,112 @@ def kinematicRouting(discharge_avg, discharge, lateral_inflow, constant, upstrea
...
             # Solve the kinematic wave for the current pixel
-            solve1Pixel(ordered_pixels[index], discharge_avg, discharge, lateral_inflow, constant, upstream_lookup,\
-                        num_upstream_pixels, a_dx_div_dt, b_a_dx_div_dt, inv_time_delta, beta, inv_beta, b_minus_1, a_dx)
+            pix = ordered_pixels[index]
+            discharge_before = discharge[pix]
+            if solve1Pixel(pix, discharge, constant[pix], upstream_lookup,\
+                        num_upstream_pixels, a_dx_div_dt[pix], b_a_dx_div_dt[pix], beta, inv_beta, b_minus_1):                        
+                solve1PixelAvg(pix, discharge_avg, discharge_before, discharge[pix], lateral_inflow[pix], upstream_lookup,\
+                            num_upstream_pixels, inv_time_delta, beta, a_dx[pix])
+
```
Comment (doc78) [inline src/lisflood/hydrological_modules/kinematic_wave_parallel_tools.py:71]: Please @ecCinziaMazzetti and @StefaniaGrimaldi just check the changes here, as I think we were using the whole arrays previously in the computation, while only the "pix" element of the a_dx, constant, lateral_inflow, a_dx_div_dt and b_a_dx_div_dt arrays were needed in the solve1pixel functions, but I can be wrong.

File: src/lisflood/hydrological_modules/kinematic_wave_parallel_tools.py [hunk 2]
```diff
@@ -60,91 +62,112 @@ def kinematicRouting(discharge_avg, discharge, lateral_inflow, constant, upstrea
...
         upstream_inflow += discharge[upstream_lookup[pix,ups_ix]]
-        upstream_inflow_avg += discharge_avg[upstream_lookup[pix, ups_ix]]
 
-    const_plus_ups_infl = upstream_inflow + constant[pix] # upstream_inflow + alpha*dx/dt*Qold**beta + dx*specific_lateral_inflow
     # If old discharge, upstream inflow and lateral inflow are below accuracy: set discharge to 0 and exit
-    if const_plus_ups_infl <= NEWTON_TOL:
+    if upstream_inflow <= NEWTON_TOL:
         discharge[pix] = 0
-        return
+        return False
```
Comment (doc78) [inline src/lisflood/hydrological_modules/kinematic_wave_parallel_tools.py:101]: @ecCinziaMazzetti @StefaniaGrimaldi Here we skip the computation of the discharge_avg value, but discharge[pix] is changed, as we set it to zero. Is that correct? Should we still need to compute channel_volume_start and channel_volume_end for dischage_avg as the quantity of discharge[pix] has changed?

**Comment by doc78** [comment]
I checked the dis.tss results of the branch master and the current branch using cold.xml. Results after the commit "split and optimization of the new solve1pixel function " are bit identical. Results after "minor optimisations" are slightly different but I think they are still OK:

29.2018 -> 29.2017
25.7968 -> 25.7967
58.7482 -> 58.7476
51.9246 -> 51.9239
...

**Comment by ecCinziaMazzetti** [comment]
Is 'minor optimisations' strictly necessary? I would very much prefer having bit identical results.

**Comment by corentincarton** [comment]
I think we can maybe do something in between that allows bit reproducibility. I think moving around some parts of the code (counts, upstream inflow, etc.) helped but maybe not the removal of closureError and the change in the while loop, which is what impacts may the bit reproducibility.

**Comment by doc78** [comment]
We run some test for speed and result differences adding the commit "[minor optimisation](https://github.com/ec-jrc/lisflood-code/pull/173/commits/73da888e317ab30c8a212df817754c611027d3dc)" and we found the following:
Speed: there is no speed difference between the two versions (with and without minor optimization).
Results differences: compared to the feature\mct_optim, the minor optimization commit introduces differences in the order of 0.000001 (absolute values), while in the version without minor optimization the differences are less then 0.0000001 (even if both version are not bit identical to the feature\mct_optim branch).
For testing, we used catchments G0629 (GloFAS3arcmin setup), 3 years of daily simulations. The simulation included only SplitRouting optional module, all other optional modules were deactivated. The comparison used the output netcdf files.

File: src/lisflood/hydrological_modules/kinematic_wave_parallel_tools.py (hunk 2 above)
Comment (corentincarton) [reply src/lisflood/hydrological_modules/kinematic_wave_parallel_tools.py:101]: @doc78, I discussed this with @ecCinziaMazzetti a few weeks ago when reviewing the changes and this looks like the correct behaviour. No need to compute channel_volume_start and channel_volume_end and discharge_avg is not computed if discharge[pix] is 0 (solve1PixelAvg not called).

**Comment by corentincarton** [comment]
@doc78, thanks for this! We did some checks on our side and we also came to the same conclusion. I think we can move forward with this version. @ecCinziaMazzetti and Nikos are currently running some performance and calibration tests, we'll share with you the results of the investigations.

**Pull request merged**